testing/unit-testing-with-bunit.md

```diff
@@ -0,0 +1,152 @@
+---
+title: Unit Testing With bUnit
+page_title: Unit Testing With bUnit
+description: Unit Test your Blazor apps by isolating the tested code from its dependencies with a mocking tool like JustMock. 
+tags: unit test, bUnit, mock, mocking, mocking tool, mocking solution, mocking software, mocking framework, Blazor
+slug: unit-test-with-bunit
+position: 5
+---
+
+# Unit Testing With bUnit
+
+This article provides information for bUnit and how it can be used with the Telerik UI for Blazor components.
+
+* [What is bUnit](#what-is-bunit)
+* [E2E Vs Unit Testing](#e2e-vs-unit-testing)
+* [BUnit Limitations](#bunit-limitations)
+* [Testing the Telerik UI for Blazor components](#testing-the-telerik-ui-for-blazor-components-with-bunit)
+* [Common issues](#common-issues)
+* [Resources](#resources)
+
+
+## What is bUnit
+
+[BUnit](https://bunit.dev/) is a popular unit testing framework for Blazor. 
+
+BUnit covers explicitly unit testing and is not designed for e2e testing. It only runs your C# and Razor code, and not JavaScript.
+
+Check the following resources for a general introduction to bUnit:
+* https://bunit.dev/docs/getting-started
+* https://youtu.be/Co7QetPYiO4
+
+## E2E Vs Unit Testing
+
+The main difference between e2e tests and unit tests is that end-to-end testing focuses on the application's behavioral flow (clicking a button that populates a message) while unit testing targets functional fragments. 
+
+End-to-end tests allow you to test the application from the user's point of view. One tool you can use for end-to-end testing in Blazor is the Selenium framework.
+
+Unit testing focuses on functional verification. You can use unit tests to ensure that the functions or calculations that generate data, a numerical value, or a text string, are working as expected.
+
+A case that requires an interaction with the component is a task for e2e testing.
+
+## BUnit Limitations
+
+A known limitation of bUnit is that it does not run JavaScript. So, if the components use some JSInterop, one should emulate IJSRuntime. You may find some more details here: https://bunit.dev/docs/test-doubles/emulating-ijsruntime.html.
+
+
+## Testing the Telerik UI for Blazor components with bUnit
+
+The Telerik UI for Blazor components rely on JSInterop to support their rich UX features. This can make testing with bUnit difficult or even impossible in some scenarios due to the above-listed limitation.
+
+You don't have to test our components, as we do that ourselves. Any unit tests for custom markup and logic have to test just that - the custom logic in an isolated scenario.
+
+Consider e2e testing for any case that targets a complete workflow.
+
+## Common issues
```

Comment:
I suggest reworking this section of the document:
* The subsections are presented in a visually unappealing way.
* The use of a bullet list isn't appropriate for this use case. Maybe a table as shown on the image below?
* Rethink the titles of the sections, especially "Dialog / Window / Any component that includes popup", "Grid - ticket 1569276", "`data-id` is different in the markup".
![image](https://github.com/telerik/blazor-docs/assets/60105689/b35b02fe-5d61-4bb1-8dc8-8a5837100240)

Reply:
I refactored the article to include tables instead of bullet lists. Only I did not include the "Problem" heading in the first column as I don't think it unifies the column contents. I left the heading cell empty instead.